src/FrequencyMapAgent.cpp

```diff
@@ -169,36 +167,52 @@ namespace geopm
         }
     }
 
+    bool FrequencyMapAgent::is_all_nan(const std::vector<double> &vec)
+    {
+        return std::all_of(vec.begin(), vec.end(),
+                           [](double x) -> bool { return std::isnan(x); });
+    }
+
     void FrequencyMapAgent::update_policy(const std::vector<double> &policy)
     {
-        m_is_policy_updated = m_freq_governor->set_frequency_bounds(
-            policy[M_POLICY_FREQ_MIN], policy[M_POLICY_FREQ_MAX]);
+        if (is_all_nan(policy)) {
+            // All-NAN policy is ignored
+            m_is_policy_updated = false;
+            return;
+        }
 
+        std::map<uint64_t, double> old_freq_map = m_hash_freq_map;
+        m_hash_freq_map.clear();
         for (auto it = policy.begin() + M_POLICY_FIRST_HASH;
-             it != policy.end() && std::next(it) != policy.end(); std::advance(it, 2)) {
+             it != policy.end() && std::next(it) != policy.end();
+             std::advance(it, 2)) {
             if (!std::isnan(*it)) {
                 auto hash = static_cast<uint64_t>(*it);
                 auto freq = *(it + 1);
-                if (std::isnan(freq)) {
-                    auto num_erased = m_hash_freq_map.erase(hash);
-                    if (num_erased != 0) {
-                        m_is_policy_updated = true;
-                    }
-                }
-                else {
-                    auto inserted = m_hash_freq_map.emplace(hash, freq);
-                    if (inserted.second) {
-                        m_is_policy_updated = true;
-                    }
-                    else {
-                        if (inserted.first->second != freq) {
-                            inserted.first->second = freq;
-                            m_is_policy_updated = true;
-                        }
-                    }
-                }
+
+                // Not valid to have NAN freq for hash.
+                // This is a logic error because it is checked by
+                // validate policy, which the controller should
+                // call before this function.
+                GEOPM_DEBUG_ASSERT(!std::isnan(freq),
+                                   "mapped region with no frequency assigned.");
```

Comment:
Shouldn't it be impossible to hit this assert?

Reply:
Yes if you are using the controller; since this is a public interface, you can still hit it through unit testing.

Reply:
That is why it is a debug assert.